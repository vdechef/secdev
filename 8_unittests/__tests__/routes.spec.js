const request = require("supertest")
const app = require("../routes.js").app
const connectMysql = require("../routes.js").connectMysql
const disconnectMysql = require("../routes.js").disconnectMysql

describe("Endpoints", () => {

    beforeEach(async () => {
        await connectMysql()
    })
    afterEach(async () => {
        await disconnectMysql()
    })

    describe("Get", () => {
        it("should return home page", async () => {
            const res = await request(app).get("/")
            expect(res.statusCode).toEqual(200)
            expect(res.body).toMatchSnapshot()
        })
        it("should return all users when no search string provided", async () => {
            const res = await request(app).get("/api/getdata?search=")
            expect(res.statusCode).toEqual(200)
            expect(res.body.length).toBe(5)
            expect(res.body).toMatchSnapshot()
        })
        it("should filter using search param", async () => {
            const res = await request(app).get("/api/getdata?search=al")
            expect(res.statusCode).toEqual(200)
            expect(res.body.length).toBe(2)
            expect(res.body).toMatchSnapshot()
        })
    })

    describe("Error handling", () => {
        it("should return generic message on SQL error", async () => {
            const res = await request(app).get("/api/getdata?search=al'")
            expect(res.statusCode).toEqual(500)
            expect(res.text).toMatchSnapshot()
        })
        it("should return dedicated message for toto", async () => {
            const res = await request(app).get("/api/getdata?search=toto")
            expect(res.statusCode).toEqual(400)
            expect(res.text).toMatchSnapshot()
        })
        it("should return dedicated message when forbidden", async () => {
            const res = await request(app).get("/api/getdata?search=bob")
            expect(res.statusCode).toEqual(500)
            expect(res.text).toMatchSnapshot()
        })
    })

})
