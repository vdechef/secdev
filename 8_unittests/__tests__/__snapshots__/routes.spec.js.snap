// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoints Error handling should return dedicated message for toto 1`] = `"Erreur valeur interdite: Toto n'a pas de baskettes"`;

exports[`Endpoints Error handling should return dedicated message when forbidden 1`] = `"Erreur inconnue"`;

exports[`Endpoints Error handling should return generic message on SQL error 1`] = `"Erreur générique pour éviter a fuite de données"`;

exports[`Endpoints Get should filter using search param 1`] = `
[
  {
    "age": 45,
    "id": 1,
    "name": "Alexandre Astier",
  },
  {
    "age": 62,
    "id": 3,
    "name": "Alain Chabat",
  },
]
`;

exports[`Endpoints Get should return all users when no search string provided 1`] = `
[
  {
    "age": 45,
    "id": 1,
    "name": "Alexandre Astier",
  },
  {
    "age": 128,
    "id": 2,
    "name": "Louis Debroglie",
  },
  {
    "age": 62,
    "id": 3,
    "name": "Alain Chabat",
  },
  {
    "age": 35,
    "id": 4,
    "name": "Lord Farquaad ",
  },
  {
    "age": 35,
    "id": 5,
    "name": "Pierre d'Argoeuvre ",
  },
]
`;

exports[`Endpoints Get should return home page 1`] = `{}`;
